src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/CpsPackageReferenceProject.cs

```diff
@@ -371,6 +371,56 @@ public override Task<string> GetCacheFilePathAsync()
             return Task.FromResult(NoOpRestoreUtilities.GetProjectCacheFilePath(cacheRoot: spec.RestoreMetadata.OutputPath));
         }
 
+        // To avoid race condition, we work on copy of cache.
+        protected override (List<FrameworkInstalledPackages> installedPackagesCopy, List<FrameworkInstalledPackages> transitivePackagesCopy) GetCacheCopy()
+        {
+            return (new List<FrameworkInstalledPackages>(InstalledPackages), new List<FrameworkInstalledPackages>(TransitivePackages));
+        }
+
+        protected override void UpdatePackageListDetails(List<FrameworkInstalledPackages> packages, IEnumerable<FrameworkInstalledPackages> detectedNewPackages)
+        {
+            if (packages == null || detectedNewPackages == null)
+            {
+                return;
+            }
+
+            var dict = new Dictionary<NuGetFramework, FrameworkInstalledPackages>();
+
+            foreach (FrameworkInstalledPackages installedPackage in packages)
+            {
+                dict[installedPackage.TargetFramework] = installedPackage;
+            }
+
+            foreach (FrameworkInstalledPackages detectedNewInstalledPackage in detectedNewPackages)
+            {
+                if (detectedNewInstalledPackage == null)
+                {
+                    continue;
+                }
+
+                FrameworkInstalledPackages frameworkPackages;
+
+                if (dict.TryGetValue(detectedNewInstalledPackage.TargetFramework, out frameworkPackages))
+                {
+                    foreach (KeyValuePair<string, ProjectInstalledPackage> package in detectedNewInstalledPackage.Packages)
+                    {
+                        if (frameworkPackages.Packages.TryGetValue(package.Key, out _))
+                        {
+                            frameworkPackages.Packages[package.Key] = package.Value;
+                        }
+                        else
+                        {
+                            frameworkPackages.Packages.Add(package.Key, package.Value);
+                        }
```

Comment:
Why do you need to check if target framework (i.e. `packageKey`) exists? After all, it seems you will replace `Packages[package.Key]` regardless of whether the key exists or not. Could it be easier if we just assign the value?:

```suggestion
                        frameworkPackages.Packages[package.Key] = package.Value;
```

Reply:
It's coming from this code path:
![image](https://user-images.githubusercontent.com/8766776/158276390-ca264a57-9310-45ea-bc9c-2d6b49754f38.png)

Yes, I thought about it. It's coded this way in below existing code, to keep logic to easy reason with existing code I decided not to change it for now.

https://github.com/NuGet/NuGet.Client/blob/fc721e9dc7f52d482f680dccad8755f0e06450ad/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/GetPackageReferenceUtility.cs#L46-L55